pages/instructorAssessmentAccess/instructorAssessmentAccess.js

```diff
@@ -10,18 +11,164 @@ const sqlLoader = require('../../prairielib/lib/sql-loader');
+};
+
+let apply_rule = function (list, formal_rule) {
+  let new_rule = Object.assign({}, formal_rule);
+  let valid = true;
+  new_rule.valid_now_public = new_rule.valid_now && (new_rule.mode_raw ?? 'Public') === 'Public';
+  new_rule.valid_now_exam = new_rule.valid_now && (new_rule.mode_raw ?? 'Exam') === 'Exam';
+  list.forEach((old_rule) => {
+    if (!valid) return;
+
+    if (old_rule.mode_raw !== null && old_rule.mode_raw !== new_rule.mode_raw) {
+      if (new_rule.mode_raw === null) {
+        // New rule is NULL, old rule is not NULL, so split new rule.
+        let new_rule_public = Object.assign({}, new_rule);
+        new_rule_public.mode_raw = 'Public';
+        new_rule_public.mode = 'Public';
+        apply_rule(list, new_rule_public);
+        let new_rule_exam = Object.assign({}, new_rule);
+        new_rule_exam.mode_raw = 'Exam';
+        new_rule_exam.mode = 'Exam';
+        apply_rule(list, new_rule_exam);
+        valid = false;
+      }
+      return;
+    }
+
+    if (old_rule.valid_now_public) {
+      new_rule.valid_now_public = false;
+    }
+    if (old_rule.valid_now_exam && old_rule.exam_uuid === new_rule.exam_uuid) {
+      new_rule.valid_now_exam = false;
+    }
+
+    // Simplification: if there are multiple rules with different
+    // CBTF exam UUIDs, they are not considered a conflict, and may all
+    // be valid simultaneously.
+    if (old_rule.exam_uuid !== new_rule.exam_uuid) return;
+
+    if (
+      compare_date(new_rule.start_date_raw, old_rule.start_date_raw, -1, +1, 0) >= 0 &&
+      compare_date(new_rule.end_date_raw, old_rule.end_date_raw, +1, -1, 0) <= 0
+    ) {
+      valid = false;
+    } else if (
+      compare_date(new_rule.start_date_raw, old_rule.start_date_raw, -1, +1, 0) >= 0 &&
+      compare_date(new_rule.start_date_raw, old_rule.end_date_raw, -1, -1, -1) <= 0
+    ) {
+      new_rule.start_date_raw = old_rule.end_date_plus_one_raw;
+      new_rule.start_date = old_rule.end_date_plus_one;
+    } else if (
+      compare_date(new_rule.end_date_raw, old_rule.start_date_raw, +1, +1, +1) >= 0 &&
+      compare_date(new_rule.end_date_raw, old_rule.end_date_raw, +1, -1, 0) <= 0
+    ) {
+      new_rule.end_date_raw = old_rule.start_date_minus_one_raw;
+      new_rule.end_date = old_rule.start_date_minus_one;
+    } else if (
+      compare_date(new_rule.start_date_raw, old_rule.start_date_raw, -1, +1, 0) < 0 &&
+      compare_date(new_rule.end_date_raw, old_rule.end_date_raw, +1, -1, 0) > 0
+    ) {
+      let rule_before = Object.assign({}, new_rule);
+      rule_before.end_date_raw = old_rule.start_date_minus_one_raw;
+      rule_before.end_date = old_rule.start_date_minus_one;
+      apply_rule(list, rule_before);
+
+      new_rule.start_date_raw = old_rule.end_date_plus_one_raw;
+      new_rule.start_date = old_rule.end_date_plus_one;
+    }
+  });
+  if (valid) {
+    list.push(new_rule);
+  }
+};
+
+router.get(
+  '/',
+  asyncHandler(async (req, res, _next) => {
```

Comment:
There's a lot of semi-complex logic here, and this kind of pure-data-in, pure-data-out thing is a fantastic candidate for a unit test.

Reply:
Done in c07247c90.